### PR TITLE
Forward implementation for LSTM operator.

### LINKED ISSUE #4629: Porting LSTM Operator
The [lstm_unit_op](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/lstm_unit_op.cc) calculates one time step for the input sequence in LSTM network. Different from it, this LSTM operator will calculate all time step for the input sequence.

- support variable-length base on LoDTensor.
- support peephole connection by default, the users also can disable it.
- support to set cell initialized state and hidden initialized  state by users.
   - The LstmLayer in PaddlePaddle now set them to zeros by default：https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/LstmLayer.cpp#L174
- support to pass the states across mini-batch.
- output cell state and hidden state.

### LINKED ISSUE #4675: SequenceToBatch
- Used in LSTM, GRU and Simple-RNN Operators.
- Related to #4629 

Maybe it also can be used in RNN group operator.

### PR DESCRIPTION
**Pull request by qingqing01**
fix #4629 
fix #4675 

- Finish the implementation kernel, but the code style will be refined later.
- Finish the forward implementation and the unit testing.
- Support reversed LSTM.
- Support peephole connections by default.
- TODO in next PR:
    - Backward implementation and it's unit testing.
    - Improve code about the kernel implementation in `math/detail`.
    - Support to disable peephole connections.
    - Enable initial hidden state and initial cell state.
    - Now use a `SequenceToBatch` functor to reorganize and sort the input sequence,  I'll try to use `TensorArray` in the future.

Directory structure:

```tree
├── lstm_op.cc
├── lstm_op.cu
├── lstm_op.h
├── math
│   ├── lstm_compute.cc
│   ├── lstm_compute.cu
│   ├── lstm_compute.h [One batch computing functor for lstm.]
│   ├── sequence2batch.cc
│   ├── sequence2batch.cu
│   ├── sequence2batch.h [Reorganize and sort the input sequence,  try to use `TensorArray` in the future.]
│   ├── detail [The code in this directory will be refined later.]
│   │   ├── CMakeLists.txt
│   │   ├── hl_activation_functions.h
│   │   ├── hl_avx_functions.cc
│   │   ├── hl_avx_functions.h
│   │   ├── hl_cpu_functions.cc
│   │   ├── hl_functions.h  [Activations for cpu, gpu and avx]
│   │   ├── hl_gpu_functions.h 
│   │   ├── lstm_cpu_kernel.h [CPU implenmentation kernel for one batch and one sequence.]
│   │   ├── lstm_gpu_kernel.h [GPU implenmentation kernel for one batch.]
│   │   └── lstm_kernel.h 
```

### CODE REVIEW

File: paddle/operators/math/sequence2batch.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    PADDLE_ENFORCE_EQ(lods.size(), 1UL, "Only support one level sequence now.");
+    auto lod = lods[0];
+
+    // Calculate the length of each sequence and
+    // sort sequence index by the length.
+    // example:  sequences = {s0, s1, s2}
+    //           s0: 0 0 0 0, s1: 1 1 1 1 1, s2: 2 2 2
+    //           seq_info[3] = {(4, 5, 1), (0, 4, 0), (9, 3, 2)}
+    //
+    struct SeqInfo {
```
Comment (reyoung) [inline paddle/operators/math/sequence2batch.h]: Move `struct SeqInfo` out of this function.

File: paddle/operators/math/sequence2batch.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    //                                5, 1, 10,
+    //                                6, 2, 11,
+    //                                7, 3,
+    //                                8}
+
+    // The batch number represents batch size after rearranging the
+    // input LodTensor. It is also the maximum length of input sequence.
+
+    paddle::framework::LoD batch_lods;
+    batch_lods.push_back(std::vector<size_t>{0});
```
Comment (reyoung) [inline paddle/operators/math/sequence2batch.h]: I am not sure, but is it working to push_back `std::vector<size_t>` in GPU.

File: paddle/operators/math/sequence2batch.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    batch_lods.push_back(std::vector<size_t>{0});
```
Comment (reyoung) [inline paddle/operators/math/sequence2batch.h]: Same as above

File: paddle/operators/math/sequence2batch.h (hunk 2 above)
Comment (reyoung) [reply paddle/operators/math/sequence2batch.h]: Maybe `emplace_back` is better.

File: paddle/operators/math/sequence2batch.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    // batch_lods[0] is the start positions for batch LoDTensor
+    int num_batch = (size_t)seq_info[0].length;
```
Comment (reyoung) [inline paddle/operators/math/sequence2batch.h]: static_cast<>

File: paddle/operators/math/detail/lstm_cpu_kernel.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+  T rState;
+  T rStateGrad;
+  T rStateAtv;
+  T rOutputGrad;
+  T rCheckI;
+  T rCheckF;
+  T rCheckO;
+  T rCheckIGrad;
+  T rCheckFGrad;
+  T rCheckOGrad;
```
Comment (luotao1) [inline paddle/operators/math/detail/lstm_cpu_kernel.h]: 同上，可以分组合并么。
下同。

File: paddle/operators/math/sequence2batch.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                  framework::LoDTensor& batch, const bool is_reverse) const {
+    auto lods = lod_tensor.lod();
+    PADDLE_ENFORCE_EQ(lods.size(), 1UL, "Only support one level sequence now.");
+    auto lod = lods[0];
+
+    // Calculate the length of each sequence and
+    // sort sequence index by the length.
+    // example:  sequences = {s0, s1, s2}
+    //           s0: 0 0 0 0, s1: 1 1 1 1 1, s2: 2 2 2
+    //           seq_info[3] = {(4, 5, 1), (0, 4, 0), (9, 3, 2)}
```
Comment (luotao1) [inline paddle/operators/math/sequence2batch.h]: 51行是怎么出来的，注释能更详细点么

File: paddle/operators/math/sequence2batch.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    for (size_t n = 0; n < num_batch; n++) {
+      int batch_id = batch_starts[n];
+      for (size_t i = 0; i < seq_info.size(); ++i) {
+        size_t seq_len = seq_info[i].length;
+        int start = seq_info[i].start;
+        if (n < seq_len) {
+          if (!is_reverse) {
+            seq2batch_idx[batch_id] = start + n;
+          } else {
+            seq2batch_idx[batch_id] = start + seq_len - 1 - n;
```
Comment (luotao1) [inline paddle/operators/math/sequence2batch.h]:
```
seq2batch_idx[batch_id]=is_reverse? start + seq_len - 1 - n: start + n;
```

**Comment by reyoung** [comment]
Please note #4952 has been merged

File: paddle/operators/lstm_op.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      // else if : support the initial hidden and cell
+
+      lstm_value.gateValue = gate_t.data<T>();
+      lstm_value.outputValue = out_t.data<T>();
+      lstm_value.stateValue = cell_t.data<T>();
+      lstm_value.stateActiveValue = cell_pre_act_t.data<T>();
+      math::LstmUnitFunctor<Place, T>::compute(ctx.device_context(), lstm_value,
+                                               frame_size, cur_batch_size,
+                                               gate_act, cell_act, cand_act);
+      lstm_value.prevStateValue = lstm_value.stateValue;
```
Comment (reyoung) [inline paddle/operators/lstm_op.h]: 这里是故意这么做的么？ lstm_value.prevStateValue永远等于 lstm_value.stateValue ?

因为lstm_value.stateValue没有被compute函数修改过。这一行如果移动到第117行之前效果也是一样的。

File: paddle/operators/math/detail/lstm_cpu_kernel.h [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+namespace paddle {
+namespace operators {
+namespace math {
+namespace detail {
+
+#ifndef __NVCC__
+
+template <class T, class Op>
+void naive_lstm_forward_one_sequence(Op op, LstmMetaValue<T> value,
```
Comment (reyoung) [inline paddle/operators/math/detail/lstm_cpu_kernel.h]: 这里是值传递，故调用这个函数的value永远不会修改。

File: paddle/operators/math/detail/lstm_kernel.h [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    state = valueIn * valueIg + prevState * valueFg;
+    valueOg = actGate(valueOg + state * checkO);
+    stateAtv = actState(state);
+    output = valueOg * stateAtv;
+  }
+#ifndef __NVCC__
+#ifndef __AVX__
+  static const bool avx = false;
+#else
+  static const bool avx = true;
```
Comment (reyoung) [inline paddle/operators/math/detail/lstm_kernel.h]: 看起来如果T是float， avx也应该等于false。 
是不是double类型在单测里没测过？

File: paddle/operators/math/lstm_compute.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+namespace paddle {
+namespace operators {
+namespace math {
+
+typedef enum {
+  HL_ACTIVATION_SIGMOID = 0,
+  HL_ACTIVATION_RELU = 1,
+  HL_ACTIVATION_TANH = 2,
+  HL_ACTIVATION_LINEAR = 3,
+  HL_ACTIVATION_END
```
Comment (reyoung) [inline paddle/operators/math/lstm_compute.h]: `HL_ACTIVATION_END` is not needed.

另外，即使需要，也叫做 `NUM_OF_ACTIVATIONS`比较好。

**Comment by reyoung** [comment]
Please review https://github.com/qingqing01/Paddle/pull/3 for several enhancements

**Comment by qingqing01** [comment]
@reyoung Thanks very much for your enhancements!

File: paddle/operators/CMakeLists.txt [hunk 12]
```diff
@@ -126,6 +127,7 @@ op_library(softmax_with_cross_entropy_op DEPS cross_entropy softmax)
 op_library(sum_op DEPS net_op)
 op_library(pool_op DEPS pooling)
 op_library(pool_with_index_op DEPS pooling)
+op_library(lstm_op DEPS sequence2batch lstm_compute math_function)
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/CMakeLists.txt]: Done.

File: paddle/operators/lstm_op.cc [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+};
+
+class LSTMOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LSTMOpMaker(framework::OpProto* proto, framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Input",
+             "(LoDTensor) the first input is a LodTensor, which support "
+             "variable-time length input sequence. The underlying tensor in "
+             "this LoDTenosr is a matrix with shape (T X 4D), where, T is the "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+             "(Tensor, optional) the initial hidden state is an optional "
+             "input. This is a tensor with shape (N x D), where N is the "
+             "batch size, D is the hidden size.");
+    AddInput("C0",
+             "(Tensor, optional) the initial cell state is an optional "
+             "input. This is a tensor with shape (N x D), where N is the "
+             "batch size. `H0` and `C0` can be NULL but only at the same time");
+    AddInput("Weight",
+             "(Tensor) the learnable hidden-hidden weights."
+             " - The shape is (D x 4*D), where D is the hidden size. "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+             " - Weight = {W_ih, W_fh, W_ch, W_oh}");
+    AddInput("Bias",
+             "(Tensor) the learnable weights, which contains two parts: "
+             "input-hidden bias weight and peephole connections weight if "
+             "seting `usePeepholes` True. "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+             "1. `usePeepholes = False` "
+             " - The shape is (1 x 4*D). "
+             " - Bias = {b_i, b_f, b_c, b_o}."
+             "2. `usePeepholes = True` "
+             " - The shape is (1 x 7*D). "
+             " - Bias = {b_i, b_f, b_c, b_o, W_ic, W_fc, W_oc}.");
+    AddOutput("BatchGate",
+              "(LoDTensor) This LoDTensor contains input gate, forget gate "
+              "and output gate aftern the nonlinear computation. This "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+              "LoDTensor has the same shape with the reorganized input, which "
+              "was also be called batch input. The LoD size is 2. The first "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+are the cell input and cell output activation functions, `tanh` is usually
+used for them. \f$\tilde{c_t}\f$ is also called candidate hidden state,
+which is computed based on the current input and the previous hidden state.
+
+Set `usePeepholes` False to disable peephole connection [2]. The formula
+is omitted here.
+
+@note These \f$W_{xi}x_{t}, W_{xf}x_{t}, W_{xc}x_{t}, W_{xo}x_{t}\f$
+operations on the input x_{t} were NOT included in this operator. The
+users can choose to use fully-connect operator before LSTM operator.
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.h [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+    math::LstmMetaValue<T> lstm_value;
+    T* bias_data = const_cast<T*>(bias->data<T>());
+    // the code styple in LstmMetaValue will be updated later.
+    lstm_value.checkIg = bias_data + 4 * frame_size;
+    lstm_value.checkFg = lstm_value.checkIg + frame_size;
+    lstm_value.checkOg = lstm_value.checkFg + frame_size;
+    lstm_value.prevStateValue = nullptr;
+
+    framework::LoDTensor batch_out;
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.h]: Done.

File: paddle/operators/math/detail/lstm_cpu_kernel.h (hunk 9 above)
Comment (qingqing01) [reply paddle/operators/math/detail/lstm_cpu_kernel.h]: 嗯，value里存的是T*,里面的值会变， 下个PR中将value本身作为引用传递吧。

File: paddle/operators/math/detail/lstm_cpu_kernel.h [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+  T rValueIg;
+  T rValueFg;
+  T rValueOg;
+  T rCheckI;
+  T rCheckF;
+  T rCheckO;
+  T rState;
+  T rPrevState = 0;
+  T rStateAtv;
+  T rOut;
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/math/detail/lstm_cpu_kernel.h]: Good suggestion.  后续要更改code style，到时候一起修改。Thanks!

File: paddle/operators/math/detail/lstm_cpu_kernel.h (hunk 5 above)
Comment (qingqing01) [reply paddle/operators/math/detail/lstm_cpu_kernel.h]: 同上。

File: paddle/operators/math/lstm_compute.h (hunk 11 above)
Comment (qingqing01) [reply paddle/operators/math/lstm_compute.h]: Thanks! 后续PR修改。

File: paddle/operators/math/sequence2batch.h (hunk 1 above)
Comment (qingqing01) [reply paddle/operators/math/sequence2batch.h]: Done by @reyoung

File: paddle/operators/math/sequence2batch.h (hunk 2 above)
Comment (qingqing01) [reply paddle/operators/math/sequence2batch.h]: 在GPU下，`push_back`/`emplace_back` 都可以正常使用，都可以直接赋给thrust::host_vector，单测里是没问题的。

File: paddle/operators/math/sequence2batch.h (hunk 7 above)
Comment (qingqing01) [reply paddle/operators/math/sequence2batch.h]: Done.

File: paddle/operators/math/detail/lstm_kernel.h (hunk 10 above)
Comment (qingqing01) [reply paddle/operators/math/detail/lstm_kernel.h]: 单测是double类型，在AVX函数调用的地方有类型判断： `std::is_same<T, float>::value`

https://github.com/PaddlePaddle/Paddle/pull/4929/files/694bc64aafa95b065ca5e2aecef3bf4116208a55#diff-3bfbafd6e96e201350bbbef605a0de03R282

```cpp
  if (Op::avx && !(frameSize & (8 - 1)) && (std::is_same<T, float>::value)) {
    avx_lstm_backward_one_sequence<T>(op, value, grad, frameSize, ...)
  }
```

**Pull request merged**